lib/HLSL/DxilValidation.cpp

```diff
@@ -5372,6 +5380,211 @@ static void ValidateEntrySignatures(ValidationContext &ValCtx) {
+    } else {
+      // other stages have no group
+      maskForGroup |= static_cast<uint32_t>(ConflictFlags::RequiresGroup);
+    }
+  }
+
+  uint32_t
+  IdentifyConflict(const DxilModule::ShaderCompatInfo &compatInfo) const {
+    uint32_t conflictMask = 0;
+
+    // Compatibility check said this shader kind is not compatible.
+    if (0 == ((1 << (uint32_t)shaderKind) & compatInfo.mask))
+      conflictMask |= (uint32_t)ConflictFlags::Stage;
+
+    // Compatibility check said this shader model is not compatible.
+    if (DXIL::CompareVersions(ValCtx.DxilMod.GetShaderModel()->GetMajor(),
+                              ValCtx.DxilMod.GetShaderModel()->GetMinor(),
+                              compatInfo.minMajor, compatInfo.minMinor) < 0)
+      conflictMask |= (uint32_t)ConflictFlags::ShaderModel;
+
+    if (compatInfo.shaderFlags.GetUsesDerivatives())
+      conflictMask |= maskForDeriv;
+
+    if (compatInfo.shaderFlags.GetRequiresGroup())
+      conflictMask |= maskForGroup;
+
+    return conflictMask;
+  }
+
+  void Diagnose(Function *F, uint32_t conflictMask, ConflictKind conflict,
+                ValidationRule rule, ArrayRef<StringRef> args = {}) {
+    if (conflictMask & (1 << (unsigned)conflict))
+      ValCtx.EmitFnFormatError(F, rule, args);
+  }
+
+  void DiagnoseConflicts(Function *F, uint32_t conflictMask) {
+    // Emit a diagnostic indicating that either the entry function or a function
+    // called by the entry function contains a disallowed operation.
+    if (F == EntryFn)
+      ValCtx.EmitFnError(EntryFn, ValidationRule::SmIncompatibleOperation);
+    else
+      ValCtx.EmitFnError(EntryFn, ValidationRule::SmIncompatibleCallInEntry);
+
+    // Emit diagnostics for each conflict found in this function.
+    Diagnose(F, conflictMask, ConflictKind::Stage,
+             ValidationRule::SmIncompatibleStage,
+             {ShaderModel::GetKindName(props.shaderKind)});
+    Diagnose(F, conflictMask, ConflictKind::ShaderModel,
+             ValidationRule::SmIncompatibleShaderModel);
+    Diagnose(F, conflictMask, ConflictKind::DerivLaunch,
+             ValidationRule::SmIncompatibleDerivLaunch,
+             {GetLaunchTypeStr(props.Node.LaunchType)});
+    Diagnose(F, conflictMask, ConflictKind::DerivThreadGroupDim,
+             ValidationRule::SmIncompatibleThreadGroupDim,
+             {std::to_string(props.numThreads[0]),
+              std::to_string(props.numThreads[1]),
+              std::to_string(props.numThreads[2])});
+    Diagnose(F, conflictMask, ConflictKind::DerivInComputeShaderModel,
+             ValidationRule::SmIncompatibleDerivInComputeShaderModel);
+    Diagnose(F, conflictMask, ConflictKind::RequiresGroup,
+             ValidationRule::SmIncompatibleRequiresGroup);
+  }
+
+  // Visit function and all functions called by it.
+  // Emit diagnostics for incompatibilities found in a function when no
+  // called functions introduced the conflict.
```

Comment:
I don't understand this sentence. It sounds like we're emitting incompatibility diagnostics when no called functions introduced the conflict? Maybe this is the intent? 
```suggestion
  // Emit diagnostics for incompatibilities found in a function when no
  // functions called by that function introduced the conflict.
  // In those cases, the called functions themselves will emit the diagnostic.
```